File: sdks/python/paycrest_sdk/gateway_client.py
"""Direct-contract off-ramp path for Python.

Design mirrors the TypeScript + Go implementations: the SDK handles
network/token metadata resolution, RSA+AES-GCM recipient encryption,
and call orchestration (allowance check, approve, createOrder). The
actual JSON-RPC signing and broadcast is delegated to a caller-supplied
``GatewayTransactor`` protocol — typically backed by ``web3.py`` but
pluggable for any web3 library the integrator already uses.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Protocol

from .encryption import build_recipient_payload, encrypt_recipient_payload
from .networks import get_network


class GatewayTransactor(Protocol):
    """Abstract signer+RPC handle the gateway path needs.

    Implementations are expected to sign and broadcast EVM transactions
    on the caller's behalf; the SDK hands them the ABI inputs rather
    than raw calldata so adapters can use high-level helpers like
    ``web3.contract.functions.createOrder(...).transact(...)``.
    """

    def chain_id(self) -> int: ...
    def sender_address(self) -> str: ...
    def allowance(self, token: str, owner: str, spender: str) -> int: ...
    def approve(self, token: str, spender: str, amount: int) -> str: ...
    def create_order(self, gateway: str, args: "GatewayCreateOrderArgs") -> str: ...


@dataclass(frozen=True)
class GatewayCreateOrderArgs:
    token: str
    amount: int
    rate: int
    sender_fee_recipient: str
    sender_fee: int
    refund_address: str
    message_hash: str


@dataclass(frozen=True)
class GatewayOrderResult:
    tx_hash: str
    approve_tx_hash: str | None
    gateway_address: str
    token_address: str
    amount: int
    rate: str
    message_hash: str
    refund_address: str
    network: str


@dataclass
class GatewayPathConfig:
    transactor: GatewayTransactor
    aggregator_public_key: str | None = None


class GatewayClient:
    def __init__(self, registry, config: GatewayPathConfig):
        self._registry = registry
        self._config = config

    def create_offramp_order(
        self,
        payload: dict,
        rate_resolver: Callable[[str, str, str, str], str],
    ) -> GatewayOrderResult:
        network = get_network(payload["source"]["network"])

        signer_chain = self._config.transactor.chain_id()
        if signer_chain and signer_chain != network.chain_id:
            raise ValueError(
                f'Transactor chainId={signer_chain} does not match network '
                f'"{network.slug}" (chainId={network.chain_id}).'
            )

        token = self._registry.get_token(network.slug, payload["source"]["currency"])

        rate = payload.get("rate")
        if not rate:
            rate = rate_resolver(
                network.slug,
                payload["source"]["currency"],
                payload["amount"],
                payload["destination"]["currency"],
            )

        recipient_input = payload["destination"]["recipient"]
        recipient = build_recipient_payload(
            institution=recipient_input["institution"],
            account_identifier=recipient_input["accountIdentifier"],
            account_name=recipient_input["accountName"],
            memo=recipient_input["memo"],
            provider_id=payload["destination"].get("providerId", ""),
        )
        public_key = self._registry.get_public_key()
        message_hash = encrypt_recipient_payload(recipient, public_key)

        amount_sub = to_subunits(payload["amount"], token.decimals)
        fee_sub = to_subunits(payload["senderFee"], token.decimals) if payload.get("senderFee") else 0
        rate_scaled = scale_rate(rate)

        refund_address = payload["source"].get("refundAddress") or self._config.transactor.sender_address()
        sender_fee_recipient = (
            payload.get("senderFeeRecipient") or "0x0000000000000000000000000000000000000000"
        )

        approve_hash = None
        needed = amount_sub + fee_sub
        if needed > 0:
            current = self._config.transactor.allowance(
                token.contract_address, self._config.transactor.sender_address(), network.gateway,
            )
            if current < needed:
                approve_hash = self._config.transactor.approve(
                    token.contract_address, network.gateway, needed,
                )

        tx_hash = self._config.transactor.create_order(
            network.gateway,
            GatewayCreateOrderArgs(
                token=token.contract_address,
                amount=amount_sub,
                rate=rate_scaled,
                sender_fee_recipient=sender_fee_recipient,
                sender_fee=fee_sub,
                refund_address=refund_address,
                message_hash=message_hash,
            ),
        )

        return GatewayOrderResult(
            tx_hash=tx_hash,
            approve_tx_hash=approve_hash,
            gateway_address=network.gateway,
            token_address=token.contract_address,
            amount=amount_sub,
            rate=rate,
            message_hash=message_hash,
            refund_address=refund_address,
            network=network.slug,
        )


def to_subunits(amount: str, decimals: int) -> int:
    """Convert a positive decimal string to integer base units."""
    import re

    trimmed = amount.strip()
    if not re.fullmatch(r"\d+(\.\d+)?", trimmed):
        raise ValueError(f'Invalid amount "{amount}"; expected positive decimal')
    parts = trimmed.split(".", 1)
    whole = parts[0]
    fraction = parts[1] if len(parts) == 2 else ""
    if len(fraction) > decimals:
        raise ValueError(
            f'Amount "{amount}" has more fractional digits than token decimals ({decimals})'
        )
    fraction = fraction.ljust(decimals, "0")
    return int(whole + fraction)


def scale_rate(rate: str) -> int:
    """Scale a decimal-string rate to the uint96 the Gateway expects."""
    import re

    trimmed = rate.strip()
    if not re.fullmatch(r"\d+(\.\d+)?", trimmed):
        raise ValueError(f'Invalid rate "{rate}"')
    parts = trimmed.split(".", 1)
    whole = parts[0]
    fraction = parts[1] if len(parts) == 2 else ""
    if len(fraction) <= 2:
        fraction = fraction.ljust(2, "0")
        return int(whole + fraction)
    value = int(whole + fraction[:2])
    if int(fraction[2]) >= 5:
        value += 1
    return value

File: sdks/go/sdk/gateway_client.go
package sdk

import (
	"context"
	"fmt"
	"math/big"
	"regexp"
	"strings"
)

// GatewayTransactor is the minimal surface the SDK needs from a
// caller-supplied web3 client (go-ethereum, rpc.Client, abigen bindings,
// anything). It keeps the core SDK dependency-free while still letting
// the "gateway" dispatch method do everything on the user's behalf:
// the SDK resolves network/token metadata, encrypts the recipient, and
// tells the transactor exactly what to approve and what createOrder
// call to broadcast.
//
// A go-ethereum-based implementation ships as a separate package under
// `sdks/go/gogeth` (see that package for a copy-pasteable adapter).
type GatewayTransactor interface {
	// ChainID returns the chain id of the signer's currently-connected
	// RPC backend. Used to guard against network/signer mismatches.
	ChainID(ctx context.Context) (*big.Int, error)
	// From returns the signer's EVM account address (lowercase hex,
	// `0x`-prefixed). Used as the default refund address.
	From() string
	// Allowance returns the ERC-20 allowance granted by `owner` to
	// `spender` on `token`.
	Allowance(ctx context.Context, token, owner, spender string) (*big.Int, error)
	// Approve broadcasts an ERC-20 approve(spender, amount) tx on
	// `token` and returns the transaction hash.
	Approve(ctx context.Context, token, spender string, amount *big.Int) (string, error)
	// CreateOrder broadcasts a Gateway.createOrder tx with the given
	// arguments and returns the transaction hash. Implementations must
	// ABI-encode the call using GatewayABI.
	CreateOrder(ctx context.Context, gateway string, args GatewayCreateOrderArgs) (string, error)
}

// GatewayCreateOrderArgs mirrors the Gateway.createOrder solidity
// signature — the transactor adapts these into ABI-encoded calldata.
type GatewayCreateOrderArgs struct {
	Token              string
	Amount             *big.Int
	Rate               *big.Int // uint96
	SenderFeeRecipient string
	SenderFee          *big.Int
	RefundAddress      string
	MessageHash        string // base64 encrypted envelope
}

// GatewayPathConfig wires the caller's transactor + optional overrides
// for the on-chain off-ramp path. Constructed via ClientOptions.Gateway.
type GatewayPathConfig struct {
	Transactor          GatewayTransactor
	AggregatorPublicKey string
}

// GatewayOrderResult is returned when CreateOfframpOrder is dispatched
// with OfframpMethodGateway.
type GatewayOrderResult struct {
	TxHash         string
	ApproveTxHash  string
	GatewayAddress string
	TokenAddress   string
	Amount         *big.Int
	Rate           string
	MessageHash    string
	RefundAddress  string
	Network        string
}

type gatewayClient struct {
	registry *aggregatorRegistry
	config   *GatewayPathConfig
}

func newGatewayClient(registry *aggregatorRegistry, config *GatewayPathConfig) (*gatewayClient, error) {
	if config == nil || config.Transactor == nil {
		return nil, fmt.Errorf("GatewayPathConfig.Transactor is required")
	}
	return &gatewayClient{registry: registry, config: config}, nil
}

func (g *gatewayClient) createOfframpOrder(
	ctx context.Context,
	payload OfframpOrderPayload,
	rateResolver func(ctx context.Context, network, token, amount, fiat string) (string, error),
) (*GatewayOrderResult, error) {
	network, err := GetNetwork(payload.Source.Network)
	if err != nil {
		return nil, err
	}

	chainID, err := g.config.Transactor.ChainID(ctx)
	if err == nil && chainID != nil && chainID.Int64() != network.ChainID {
		return nil, fmt.Errorf("transactor chainId=%d does not match network %q (chainId=%d)",
			chainID.Int64(), network.Slug, network.ChainID)
	}

	token, err := g.registry.getToken(ctx, network.Slug, payload.Source.Currency)
	if err != nil {
		return nil, err
	}

	rate := payload.Rate
	if rate == "" {
		rate, err = rateResolver(ctx, network.Slug, payload.Source.Currency, payload.Amount, payload.Destination.Currency)
		if err != nil {
			return nil, err
		}
	}

	recipient, err := BuildRecipientPayload(
		payload.Destination.Recipient.Institution,
		payload.Destination.Recipient.AccountIdentifier,
		payload.Destination.Recipient.AccountName,
		payload.Destination.Recipient.Memo,
		payload.Destination.ProviderID,
		nil,
	)
	if err != nil {
		return nil, err
	}
	publicKey, err := g.registry.getPublicKey(ctx)
	if err != nil {
		return nil, err
	}
	messageHash, err := EncryptRecipientPayload(recipient, publicKey)
	if err != nil {
		return nil, err
	}

	amountSub, err := ToSubunits(payload.Amount, token.Decimals)
	if err != nil {
		return nil, err
	}
	feeSub := big.NewInt(0)
	if payload.SenderFee != "" {
		feeSub, err = ToSubunits(payload.SenderFee, token.Decimals)
		if err != nil {
			return nil, err
		}
	}
	rateUint96, err := ScaleRate(rate)
	if err != nil {
		return nil, err
	}

	refundAddr := g.config.Transactor.From()
	if payload.Source.RefundAddress != "" {
		refundAddr = payload.Source.RefundAddress
	}
	senderFeeRecipient := payload.SenderFeeRecipient
	if senderFeeRecipient == "" {
		senderFeeRecipient = "0x0000000000000000000000000000000000000000"
	}

	// 1) Approve gateway if current allowance is insufficient.
	needed := new(big.Int).Add(amountSub, feeSub)
	var approveHash string
	if needed.Sign() > 0 {
		current, err := g.config.Transactor.Allowance(ctx, token.ContractAddress, g.config.Transactor.From(), network.Gateway)
		if err != nil {
			return nil, fmt.Errorf("read allowance: %w", err)
		}
		if current.Cmp(needed) < 0 {
			approveHash, err = g.config.Transactor.Approve(ctx, token.ContractAddress, network.Gateway, needed)
			if err != nil {
				return nil, fmt.Errorf("approve: %w", err)
			}
		}
	}

	// 2) Gateway.createOrder
	createHash, err := g.config.Transactor.CreateOrder(ctx, network.Gateway, GatewayCreateOrderArgs{
		Token:              token.ContractAddress,
		Amount:             amountSub,
		Rate:               rateUint96,
		SenderFeeRecipient: senderFeeRecipient,
		SenderFee:          feeSub,
		RefundAddress:      refundAddr,
		MessageHash:        messageHash,
	})
	if err != nil {
		return nil, fmt.Errorf("createOrder: %w", err)
	}

	return &GatewayOrderResult{
		TxHash:         createHash,
		ApproveTxHash:  approveHash,
		GatewayAddress: network.Gateway,
		TokenAddress:   token.ContractAddress,
		Amount:         amountSub,
		Rate:           rate,
		MessageHash:    messageHash,
		RefundAddress:  refundAddr,
		Network:        network.Slug,
	}, nil
}

var decimalRegex = regexp.MustCompile(`^\d+(\.\d+)?$`)

// ToSubunits converts a positive decimal string amount into integer
// base units (e.g. "1.5" with decimals=6 -> 1_500_000).
func ToSubunits(amount string, decimals int) (*big.Int, error) {
	trimmed := strings.TrimSpace(amount)
	if trimmed == "" {
		return nil, fmt.Errorf("empty amount")
	}
	if !decimalRegex.MatchString(trimmed) {
		return nil, fmt.Errorf("invalid amount %q; expected positive decimal", amount)
	}
	parts := strings.SplitN(trimmed, ".", 2)
	whole := parts[0]
	fraction := ""
	if len(parts) == 2 {
		fraction = parts[1]
	}
	if len(fraction) > decimals {
		return nil, fmt.Errorf("amount %q has more fractional digits than token decimals (%d)", amount, decimals)
	}
	fraction = fraction + strings.Repeat("0", decimals-len(fraction))
	out, ok := new(big.Int).SetString(whole+fraction, 10)
	if !ok {
		return nil, fmt.Errorf("invalid integer conversion for %q", amount)
	}
	return out, nil
}

// ScaleRate converts a decimal-string rate to the uint96 representation
// the Gateway contract expects (rate * 100 rounded to nearest).
func ScaleRate(rate string) (*big.Int, error) {
	trimmed := strings.TrimSpace(rate)
	if !decimalRegex.MatchString(trimmed) {
		return nil, fmt.Errorf("invalid rate %q", rate)
	}
	parts := strings.SplitN(trimmed, ".", 2)
	whole := parts[0]
	fraction := ""
	if len(parts) == 2 {
		fraction = parts[1]
	}
	if len(fraction) <= 2 {
		fraction = fraction + strings.Repeat("0", 2-len(fraction))
		out, ok := new(big.Int).SetString(whole+fraction, 10)
		if !ok {
			return nil, fmt.Errorf("invalid rate integer conversion for %q", rate)
		}
		return out, nil
	}
	shifted := whole + fraction[:2]
	out, ok := new(big.Int).SetString(shifted, 10)
	if !ok {
		return nil, fmt.Errorf("invalid rate integer conversion for %q", rate)
	}
	if fraction[2]-'0' >= 5 {
		out = out.Add(out, big.NewInt(1))
	}
	return out, nil
}
